server/src/main/java/org/opensearch/extensions/ExtensionRequest.java

```diff
@@ -24,41 +25,60 @@
  */
 public class ExtensionRequest extends TransportRequest {
     private static final Logger logger = LogManager.getLogger(ExtensionRequest.class);
-    private ExtensionsManager.RequestType requestType;
+    private final ExtensionsManager.RequestType requestType;
+    private final String uniqueId;
 
     public ExtensionRequest(ExtensionsManager.RequestType requestType) {
+        this(requestType, null);
+    }
+
+    public ExtensionRequest(ExtensionsManager.RequestType requestType, String uniqueId) {
         this.requestType = requestType;
+        this.uniqueId = uniqueId;
     }
 
     public ExtensionRequest(StreamInput in) throws IOException {
         super(in);
         this.requestType = in.readEnum(ExtensionsManager.RequestType.class);
+        this.uniqueId = in.readOptionalString();
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         out.writeEnum(requestType);
+        out.writeOptionalString(uniqueId);
     }
 
     public ExtensionsManager.RequestType getRequestType() {
         return this.requestType;
     }
 
+    @Nullable
+    public String getUniqueId() {
+        return uniqueId;
+    }
+
     public String toString() {
-        return "ExtensionRequest{" + "requestType=" + requestType + '}';
+        return "ExtensionRequest{" + "requestType=" + requestType + "uniqueId=" + uniqueId + '}';
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         ExtensionRequest that = (ExtensionRequest) o;
-        return Objects.equals(requestType, that.requestType);
+        if (uniqueId == null) {
+            return Objects.equals(requestType, that.requestType) && true;
+        } else if (that.uniqueId == null) {
+            return Objects.equals(requestType, that.requestType) && false;
+        } else {
+            return Objects.equals(requestType, that.requestType) && uniqueId.equals(that.uniqueId);
```

Comment:
Just for my understanding why are we checking uniqueId as null here? Should have been checked before

Reply:
Same as next comment, uniqueId will have possible to be `null` here.  we should need to check uniqueId as null here.

Reply:
I agree that storing `uniqueId` as a possible `null` creates a requirement to check it elsewhere.   If we handle the input stream reading and make it an optional there (instead of null) then we could rely on Java to handle `.equals()` between two optionals without an extra check.